File: 02.每日一题/464.我能赢吗/canIWin.cpp
//
// Created by lifan on 2022/5/22.
//
#include <unordered_map>
#include <iostream>

class Solution {
  std::unordered_map<int, bool> mem;

 private:
  bool df(int usedNums, int currentTotal, int maxChoosableInteger, int desiredTotal) {
      if (mem.count(usedNums) == 0) {
          bool res = false;
          for (int i = 0; i < maxChoosableInteger; ++i) {
              if (((usedNums >> i) & 1) == 0) {
                  if (currentTotal + i + 1 >= desiredTotal ||
                      !df(usedNums | 1 << i, currentTotal + i + 1, maxChoosableInteger, desiredTotal)) {
                      res = true;
                      break;
                  }
              }
          }
          mem[usedNums] = res;
      }

      return mem[usedNums];
  }

 public:
  bool canIWin(int maxChoosableInteger, int desiredTotal) {
      //如果所有数加起来都比期望小
      if ((1 + maxChoosableInteger) * maxChoosableInteger / 2 < desiredTotal) {
              return false;
      }

      return df(0, 0, maxChoosableInteger, desiredTotal);
  }
};

int main() {
    Solution *s = new Solution();
    std::cout << s->canIWin(4, 6);
}

File: 02.每日一题/464.我能赢吗/canIWin.py
from functools import cache


class Solution:
    def canIWin(self, maxChoosableInteger: int, desiredTotal: int) -> bool:
        @cache  # 开启缓存
        def df(usedNums: int, currentTotal: int) -> bool:
            for i in range(maxChoosableInteger):
                if (usedNums >> i) & 1 == 0:
                    if currentTotal + i + 1 >= desiredTotal or not df(usedNums | 1 << i, currentTotal + i + 1):
                        return True
            return False

        return (1 + maxChoosableInteger) * maxChoosableInteger // 2 >= desiredTotal and df(0, 0)


if __name__ == "__main__":
    s = Solution()
    print(s.canIWin(10, 11))
